mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala

```diff
@@ -0,0 +1,466 @@
+  }
+
+  /**
+   * Add a new training instance to this HuberAggregator, and update the loss and gradient
+   * of the objective function.
+   *
+   * @param instance The instance of data point to be added.
+   * @return This HuberAggregator object.
+   */
+  def add(instance: Instance): this.type = {
+    instance match { case Instance(label, weight, features) =>
+      require(dim == features.size, s"Dimensions mismatch when adding new sample." +
+        s" Expecting $dim but got ${features.size}.")
+      require(weight >= 0.0, s"instance weight, $weight has to be >= 0.0")
+
+      if (weight == 0.0) return this
+
+      val margin = {
+        var sum = 0.0
+        features.foreachActive { (index, value) =>
+          if (featuresStd(index) != 0.0 && value != 0.0) {
+            sum += coefficients(index) * (value / featuresStd(index))
+          }
+        }
+        sum + intercept
+      }
+      val linearLoss = label - margin
+
+      if (math.abs(linearLoss) <= sigma * m) {
+        lossSum += weight * (sigma +  math.pow(linearLoss, 2.0) / sigma)
+
+        gradientSumArray(0) += weight * (1.0 - math.pow(linearLoss / sigma, 2.0))
+        if (fitIntercept) {
+          gradientSumArray(1) += weight * -2.0 * linearLoss / sigma
+        }
+        features.foreachActive { (index, value) =>
+          if (featuresStd(index) != 0.0 && value != 0.0) {
+            gradientSumArray(index + 2) +=
+              weight * -2.0 * linearLoss / sigma * (value / featuresStd(index))
+          }
+        }
+      } else {
+        val sign = if (linearLoss >= 0) -1.0 else 1.0
+        lossSum += weight * (sigma + 2.0 * m * math.abs(linearLoss) - sigma * m * m)
+        gradientSumArray(0) += weight * (1.0 - m * m)
+        if (fitIntercept) {
+          gradientSumArray(1) += weight * (sign * 2.0 * m)
+        }
+        features.foreachActive { (index, value) =>
+          if (featuresStd(index) != 0.0 && value != 0.0) {
+            gradientSumArray(index + 2) += weight * sign * 2.0 * m * (value / featuresStd(index))
+          }
+        }
+      }
+
+      totalCnt += 1
+      weightSum += weight
+      this
+    }
+  }
+
+  /**
+   * Merge another HuberAggregator, and update the loss and gradient
+   * of the objective function.
+   * (Note that it's in place merging; as a result, `this` object will be modified.)
+   *
+   * @param other The other HuberAggregator to be merged.
+   * @return This HuberAggregator object.
+   */
+  def merge(other: HuberAggregator): this.type = {
+    require(dim == other.dim, s"Dimensions mismatch when merging with another " +
+      s"HuberAggregator. Expecting $dim but got ${other.dim}.")
+
+    if (other.weightSum != 0) {
+      totalCnt += other.totalCnt
+      weightSum += other.weightSum
+      lossSum += other.lossSum
+
+      var i = 0
+      val localThisGradientSumArray = this.gradientSumArray
+      val localOtherGradientSumArray = other.gradientSumArray
+      while (i < gradientSumArray.length) {
```

Comment:
Although this will be called far less than the `add` method, I believe we still consider this performance critical. A while loop will be faster and require no virtual function calls compared with a `foreach`. We don't use `map` because we are modifying in place here.